source/mqtt5_client.c

```diff
@@ -0,0 +1,2041 @@
+        disconnect ? Py_True : Py_False,                                                  /* O */
+        disconnect ? disconnect->reason_code : 0,                                         /* i */
+        (disconnect && disconnect->session_expiry_interval_seconds) ? Py_True : Py_False, /* O */
+        (disconnect && disconnect->session_expiry_interval_seconds) ? *disconnect->session_expiry_interval_seconds
+                                                                    : 0,                   /* k */
+        (disconnect && disconnect->reason_string) ? disconnect->reason_string->ptr : NULL, /* s# */
+        (disconnect && disconnect->reason_string) ? disconnect->reason_string->len : 0,
+        user_property_count > 0 ? user_properties_list : Py_None,                                /* O */
+        (disconnect && disconnect->server_reference) ? disconnect->server_reference->ptr : NULL, /* s# */
+        (disconnect && disconnect->server_reference) ? disconnect->server_reference->len : 0);
+    if (!result) {
+        PyErr_WriteUnraisable(PyErr_Occurred());
+    }
+cleanup:
+    Py_XDECREF(result);
+    Py_XDECREF(user_properties_list);
+
+    PyGILState_Release(state);
+}
+
+static void s_lifecycle_event_callback(const struct aws_mqtt5_client_lifecycle_event *event) {
+    switch (event->event_type) {
+        case AWS_MQTT5_CLET_STOPPED:
+            s_lifecycle_event_stopped(event->user_data);
+            break;
+
+        case AWS_MQTT5_CLET_ATTEMPTING_CONNECT:
+            s_lifecycle_event_attempting_connect(event->user_data);
+            break;
+
+        case AWS_MQTT5_CLET_CONNECTION_SUCCESS:
+            s_lifecycle_event_connection_success(event->user_data, event->connack_data, event->settings);
+            break;
+
+        case AWS_MQTT5_CLET_CONNECTION_FAILURE:
+            s_lifecycle_event_connection_failure(event->user_data, event->connack_data, event->error_code);
+            break;
+
+        case AWS_MQTT5_CLET_DISCONNECTION:
+            s_lifecycle_event_disconnection(event->user_data, event->disconnect_data, event->error_code);
+            break;
+
+        default:
+            break;
+    }
+}
+
+/*******************************************************************************
+ * Websocket
+ ******************************************************************************/
+
+/* persistent storage for the data needed to transform the websocket handshake*/
+struct ws_handshake_transform_data {
+    /* Strong reference to Python Client so it can't go out of scope during transform */
+    PyObject *client_core;
+
+    struct aws_http_message *request;
+    aws_mqtt5_transform_websocket_handshake_complete_fn *complete_fn;
+    void *complete_ctx;
+
+    /* Python bindings we created to wrap the native request */
+    /* Necessary? */
+    PyObject *request_binding_py;
+    PyObject *headers_binding_py;
+};
+
+static const char *s_capsule_name_ws_handshake_transform_data = "aws_ws_handshake_transform_data";
```

Comment:
Extremely minor: Should this be with the other static constants near the top of the file?